adapter/service/parameters/model.go

```diff
@@ -0,0 +1,381 @@
+package parameters
+
+import (
+	"fmt"
+)
+
+// ValueType represents a value type.
+type ValueType string
+
+// Constants below represent allowed value types.
+const (
+	ValueTypeInt         ValueType = "int"
+	ValueTypeString      ValueType = "string"
+	ValueTypeBool        ValueType = "bool"
+	ValueTypeIntArray    ValueType = "int_array"
+	ValueTypeStringArray ValueType = "string_array"
+)
+
+func AllowedValueTypes() []ValueType {
+	return []ValueType{
+		ValueTypeInt,
+		ValueTypeString,
+		ValueTypeBool,
+		ValueTypeIntArray,
+		ValueTypeStringArray,
+	}
+}
+
+func IsValueTypeAllowed(t ValueType) bool {
+	for _, allowed := range AllowedValueTypes() {
+		if t == allowed {
+			return true
+		}
+	}
+
+	return false
+}
+
+// WidgetType represents a widget type.
+type WidgetType string
+
+// Constants below represent allowed widget types.
+const (
+	WidgetTypeInput       WidgetType = "input"
+	WidgetTypeSelect      WidgetType = "select"
+	WidgetTypeMultiSelect WidgetType = "multiselect"
+)
+
+var ( //nolint:gofumpt
+	// widgetTypeToValueTypeMapping maps widget types to allowed value types.
+	widgetTypeToValueTypeMapping = map[WidgetType][]ValueType{
+		WidgetTypeInput:       {ValueTypeInt, ValueTypeString, ValueTypeBool},
+		WidgetTypeSelect:      {ValueTypeInt, ValueTypeString},
+		WidgetTypeMultiSelect: {ValueTypeIntArray, ValueTypeStringArray},
+	}
+)
+
+// ParameterSpecification represents a parameter specification that must be provided by the Controller.
+type ParameterSpecification struct {
+	ID           string        `json:"parameter_id"`
+	Name         string        `json:"name"`
+	Description  string        `json:"description"`
+	ValueType    ValueType     `json:"value_type"`
+	WidgetType   WidgetType    `json:"widget_type"`
+	Options      SelectOptions `json:"options,omitempty"`
+	Min          int           `json:"min,omitempty"`
+	Max          int           `json:"max,omitempty"`
+	DefaultValue any           `json:"default_value"`
+	ReadOnly     bool          `json:"read_only"`
+}
+
+// ValidateParameter validates a parameter against the specification.
+func (s ParameterSpecification) ValidateParameter(p Parameter) error {
+	if err := s.validateTypeMatching(p); err != nil {
+		return err
+	}
+
+	switch s.WidgetType {
+	case WidgetTypeInput:
+		return s.validateInput(p)
+	case WidgetTypeSelect:
+		return s.validateSelect(p)
+	case WidgetTypeMultiSelect:
+		return s.validateMultiSelect(p)
+	default:
+		return fmt.Errorf("widget type '%s' is not supported", s.WidgetType)
+	}
+}
+
+func (s ParameterSpecification) validateTypeMatching(p Parameter) error {
```

Comment:
This is not needed. A ValueType is defined explicitly on ParametersSpecification.ValueType

You don't need to infer anything. Specifications tells you what is the accepted ValueType. You just compare incoming param type with that of spec. If the widget was multiselect and the value type in spec was string array, but i'd send int array, it would pass your validation, but would be incorrect.

Reply:
You're right, fixed.